xcode/Subconscious/Shared/Components/Common/Forms/ValidatedTextField.swift

```diff
@@ -7,17 +7,56 @@
 
 import SwiftUI
 import Combine
+import ObservableStore
+
+struct ValidatedFormField<T: Equatable, Action>: View {
+    var alignment: HorizontalAlignment = .leading
+    var placeholder: String
+    var field: FormField<String, T>
+    var send: (Action) -> Void
+    var tag: (FormFieldAction<String>) -> Action
+    var caption: String
+    var axis: Axis = .horizontal
+    var autoFocus: Bool = false
+    var submitLabel: SubmitLabel = .done
+    var onSubmit: () -> Void = {}
+    
+    
+    var body: some View {
+        ValidatedTextField(
+            alignment: alignment,
+            placeholder: placeholder,
+            text: Binding(
+                get: { field.value },
+                send: send,
+                tag: { v in tag(.setValue(input: v))}
+            ),
+            onFocusChanged: { focused in
+                send(tag(.focusChange(focused: focused)))
+            },
+            caption: caption,
+            axis: axis,
+            autoFocus: autoFocus,
+            isValid: !field.shouldPresentAsInvalid,
+            submitLabel: submitLabel,
+            onSubmit: onSubmit
+        )
+        .formField()
+    }
+}
 
 /// A text field that comes with help text and a validation flag
 struct ValidatedTextField: View {
+    @State var innerText: String = ""
+    
     var alignment: HorizontalAlignment = .leading
     var placeholder: String
     @Binding var text: String
     var onFocusChanged: (Bool) -> Void = { _ in}
     var caption: String
-    var hasError: Bool = false
     var axis: Axis = .horizontal
     var autoFocus: Bool = false
+    var isValid: Bool = true
```

Comment:
I changed this back to `isValid` and renamed the concept of `hasError` to `shouldPresentAsInvalid` in `FormField`.